src/devices/Arduino/ArduinoGpioControllerDriver.cs

```diff
@@ -0,0 +1,305 @@
+                    throw new NotSupportedException($"Mode {mode} is not supported for this operation");
+            }
+
+            var pinConfig = _supportedPinConfigurations.FirstOrDefault(x => x.Pin == pinNumber);
+            if (pinConfig == null || !pinConfig.PinModes.Contains(firmataMode))
+            {
+                throw new NotSupportedException($"Mode {mode} is not supported on Pin {pinNumber}.");
+            }
+
+            _arduinoBoard.Firmata.SetPinMode(pinNumber, firmataMode);
+
+            // Cache this value. Since the GpioController calls GetPinMode when trying to write a pin (to verify whether it is set to output),
+            // that would be very expensive here. And setting output pins should be cheap.
+            _pinModes[pinNumber] = mode;
+        }
+
+        protected override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinModes.TryGetValue(pinNumber, out var existingValue))
+            {
+                return existingValue;
+            }
+
+            SupportedMode mode = _arduinoBoard.Firmata.GetPinMode(pinNumber);
+
+            PinMode ret;
+            switch (mode)
+            {
+                case SupportedMode.DigitalOutput:
+                    ret = PinMode.Output;
+                    break;
+                case SupportedMode.InputPullup:
+                    ret = PinMode.InputPullUp;
+                    break;
+                case SupportedMode.DigitalInput:
+                    ret = PinMode.Input;
+                    break;
+                default:
+                    ret = PinMode.Input; // TODO: Return "Unknown"
+                    break;
+            }
+
+            _pinModes[pinNumber] = ret;
+            return ret;
+        }
+
+        protected override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            SupportedMode firmataMode;
+            switch (mode)
+            {
+                case PinMode.Output:
+                    firmataMode = SupportedMode.DigitalOutput;
+                    break;
+                case PinMode.InputPullUp:
+                    firmataMode = SupportedMode.InputPullup;
+                    break;
+                case PinMode.Input:
+                    firmataMode = SupportedMode.DigitalInput;
+                    break;
+                default:
+                    return false;
+            }
+
+            var pinConfig = _supportedPinConfigurations.FirstOrDefault(x => x.Pin == pinNumber);
+            if (pinConfig == null || !pinConfig.PinModes.Contains(firmataMode))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        protected override PinValue Read(int pinNumber)
+        {
+            return _arduinoBoard.Firmata.GetDigitalPinState(pinNumber);
+        }
+
+        protected override void Write(int pinNumber, PinValue value)
+        {
+            _arduinoBoard.Firmata.WriteDigitalPin(pinNumber, value);
```

Comment:
nit: I find the naming on the `Firmata` really weird, consider: `ReadDigitalPin`/`WriteDigitalPin`.